File: src/main/java/com/temelyan/pomoapp/util/ExceptionHandler/ExceptionInfoHandler.java
package com.temelyan.pomoapp.util.ExceptionHandler;

import com.temelyan.pomoapp.util.ValidationUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import javax.servlet.http.HttpServletRequest;

@RestControllerAdvice(annotations = RestController.class)
@Order(Ordered.HIGHEST_PRECEDENCE + 5)
public class ExceptionInfoHandler {
    private Logger logger = LoggerFactory.getLogger(getClass());

    @ResponseStatus(value = HttpStatus.CONFLICT)  // 409
    @ExceptionHandler(DataIntegrityViolationException.class)
    public ErrorInfo conflict(HttpServletRequest req, DataIntegrityViolationException e) {
        String rootMsg = ValidationUtil.getRootCause(e).getMessage();
        if (rootMsg != null) {
            if (rootMsg.toLowerCase().contains("project_unique_name_user")) {
                return logAndGetErrorInfo(req, e, false, "Data Error", "project with this username already exists");
            }
        }
        return logAndGetErrorInfo(req, e, true, "Data error");
    }

    private ErrorInfo logAndGetErrorInfo(HttpServletRequest req, Exception e, boolean logException, String errorType, String... details) {
        Throwable rootCause = ValidationUtil.getRootCause(e);
        if (logException) {
            logger.error(errorType + " at request " + req.getRequestURL(), rootCause);
        } else {
            logger.warn("{} at request  {}: {}", errorType, req.getRequestURL(), rootCause.toString());
        }
        return new ErrorInfo(req.getRequestURL(), errorType,
                "Data error",
                details.length != 0 ? details : new String[]{rootCause.toString()});
    }
}

File: src/main/java/com/temelyan/pomoapp/util/ExceptionHandler/ExceptionInfoHandler.kt
package com.temelyan.pomoapp.util.ExceptionHandler

import com.temelyan.pomoapp.util.ValidationUtil
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RestControllerAdvice

import javax.servlet.http.HttpServletRequest

@RestControllerAdvice(annotations = [RestController::class])
@Order(Ordered.HIGHEST_PRECEDENCE + 5)
class ExceptionInfoHandler {
    private val logger = LoggerFactory.getLogger(javaClass)

    @ResponseStatus(value = HttpStatus.CONFLICT)  // 409
    @ExceptionHandler(DataIntegrityViolationException::class)
    fun conflict(req: HttpServletRequest, e: DataIntegrityViolationException): ErrorInfo {
        val rootMsg = ValidationUtil.getRootCause(e).message
        if (rootMsg != null) {
            if (rootMsg.toLowerCase().contains("project_unique_name_user")) {
                return logAndGetErrorInfo(req, e, false, "Data Error", "project with this username already exists")
            }
        }
        return logAndGetErrorInfo(req, e, true, "Data error")
    }

    private fun logAndGetErrorInfo(req: HttpServletRequest, e: Exception, logException: Boolean, errorType: String, vararg details: String): ErrorInfo {
        val rootCause = ValidationUtil.getRootCause(e)
        if (logException) {
            logger.error(errorType + " at request " + req.requestURL, rootCause)
        } else {
            logger.warn("{} at request  {}: {}", errorType, req.requestURL, rootCause.toString())
        }
        return ErrorInfo(req.requestURL, errorType,
                "Data error",
                *if (details.size != 0) details else arrayOf(rootCause.toString()))
    }
}
